src/ens/ens_fs.c

```diff
@@ -0,0 +1,144 @@
+    fs->entry_size = entry_size;
+
+    // get all information needed for the needed for the fs
+    const struct device* dev = flash_area_get_device(fs->area);
+    fs->sector_count = flash_get_page_count(dev);
+    struct flash_pages_info info;
+    if (flash_get_page_info_by_offs(dev, fs->area->fa_off, &info)) {
+        // on error, close the flash area
+        flash_area_close(fs->area);
+        return -ENS_INTERR;
+    }
+    fs->sector_size = info.size;
+
+    k_mutex_init(&fs->ens_fs_lock);
+    return 0;
+}
+
+int ens_fs_read(ens_fs_t* fs, uint64_t id, void* dest) {
+    int rc = 0;
+    k_mutex_lock(&fs->ens_fs_lock, K_FOREVER);
+
+    // read the entry from flash
+    uint64_t offset = id * fs->entry_size;
+
+    if (flash_area_read(fs->area, offset, dest, fs->entry_size)) {
+        // opening of flash area was not successful
+        rc = -ENS_INTERR;
+        goto end;
+    }
+
+    // check, if checksum and not-deleted flag are as expected
+    uint8_t* obj = dest;
+
+    int isNotDeleted = obj[fs->entry_size - 1] & 1;
+    if (!isNotDeleted) {
+        // entry got deleted
+        rc = -ENS_DELENT;
+        goto end;
+    }
+
+    // crc stored in entry
+    uint8_t entryCRC = GET_CHECKSUM(obj[fs->entry_size - 1]);
+
+    // calculated crc
+    uint8_t checkCRC = crc7_be(SEED, obj, fs->entry_size - 1);
+
+    int isInvalid = memcmp(&entryCRC, &checkCRC, 1);
+    if (isInvalid) {
+        // if checksum is not equal to calculated checksum or if deleted flag is not 1, set memory to 0
+        memset(dest, 0, fs->entry_size);
+        // we do not know, if object got deleted or data is corrupt
+        rc = -ENS_NOENT;
+    }
+
+end:
+    k_mutex_unlock(&fs->ens_fs_lock);
+    return rc;
+}
+
+int ens_fs_write(ens_fs_t* fs, uint64_t id, void* data) {
+    int rc = 0;
+    k_mutex_lock(&fs->ens_fs_lock, K_FOREVER);
+
+    // set CRC and not-deleted-flag
+    uint8_t* obj = data;
+    obj[fs->entry_size - 1] = crc7_be(SEED, obj, fs->entry_size - 1) | 1;
+
+    uint64_t offset = id * fs->entry_size;
+    if (flash_area_write(fs->area, offset, data, fs->entry_size)) {
+        // writing to flash was not successful
+        rc = -ENS_INTERR;
+    }
+
+    k_mutex_unlock(&fs->ens_fs_lock);
+    return rc;
+}
+
+int ens_fs_delete(ens_fs_t* fs, uint64_t id) {
+    int rc = 0;
+    k_mutex_lock(&fs->ens_fs_lock, K_FOREVER);
+
+    uint8_t data[fs->entry_size];
+
+    uint64_t offset = id * fs->entry_size;
+    if (flash_area_read(fs->area, offset, data, fs->entry_size)) {
+        // reading was not successful
+        rc = -ENS_INTERR;
+        goto end;
+    }
+
+    // set memory to 0, so not-deleted flag is 0
+    memset(data, 0, fs->entry_size);
+    if (flash_area_write(fs->area, offset, data, fs->entry_size)) {
+        // writing was not successful
+        rc = -ENS_INTERR;
+    }
+
+end:
+    k_mutex_unlock(&fs->ens_fs_lock);
+    return rc;
+}
+
+int ens_fs_page_erase(ens_fs_t* fs, uint64_t entry_id, uint64_t sector_count) {
```

Comment:
```suggestion
// Returns the amount of theoretically deleted entries (e.g. entries, entry_id up to entry_id+ret could be deleted)
uint64_t ens_fs_make_space(ens_fs_t* fs, uint64_t entry_id) {
```

Reply:
The function should return an error if entry_id is not at the start of the page.